### PR TITLE
Issue #24 - Initial X-HM:// URI and QR Setup Code work

### PR DESCRIPTION
**Pull request by jslay88**
For Issue #24 

Needs further discussion, such as how to make the bitwise ops more dynamic. 

My lack of proper Apple Documentation limits this for me. :)

### CODE REVIEW

**Comment by ikalchev** [comment]
Hey thanks for contributing! I am currently at work, so I will have deeper look in a few hours and go through Apple's spec (I haven't read it since it went out "public" a few months ago, but I do have a copy now).

But it looks very nice, good work. Will try to merge today.

**Comment by jslay88** [comment]
Let me know what changes I need to make, or you can make them yourself. Just trying to get more openly involved in projects, so I'm new to the process.

**Comment by ikalchev** [comment]
Sure no worries - if something needs changing I will tell you right away to add it. I am also new to the process.

**Comment by jslay88** [comment]
I built the HAP-NodeJS method out in Python, but it doesn't appear to work. Maybe I am missing something.

**Comment by jslay88** [comment]
Ok, should be good to go now.

File: pyhap/accessory.py [hunk 1]
```diff
@@ -143,6 +147,10 @@ def __init__(self, display_name, aid=None, mac=None, pincode=None,
         self.config_version = 2
         self.reachable = True
         self.pincode = pincode
+        if setup_id is None:
```
Comment (ikalchev) [inline pyhap/accessory.py]: The if-else i one line: `self._setup_id = setup_id or self._generate_setup_id()`. However, if you think the if-else is more readable, I don't mind.

File: pyhap/accessory.py [hunk 2]
```diff
@@ -162,6 +170,15 @@ def __getstate__(self):
         state["run_sentinel"] = None
         return state
 
+    @staticmethod
+    def _generate_setup_id():
+        chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+        rand_bytes = urandom(8)
+        setup_id = ''
+        for x in range(4):
+            setup_id += chars[struct.unpack_from('i', rand_bytes, x)[0] % 26]
```
Comment (ikalchev) [inline pyhap/accessory.py]: Shouldn't this be %36 (length of chars)?

File: pyhap/accessory.py [hunk 3]
```diff
@@ -273,6 +290,43 @@ def remove_paired_client(self, client_uuid):
...
+        """Generates the X-HM:// uri (Setup Code URI)
+        @rtype: str
+        """
+        buffer = bytearray(b'\x00\x00\x00\x00\x00\x00\x00\x00')
+
+        value_low = int(self.pincode.replace(b'-', b''), 10)
+        value_low |= 1 << 28
+        struct.pack_into('>L', buffer, 4, value_low)
+
+        if self.category & 1:
```
Comment (ikalchev) [inline pyhap/accessory.py]: Could you expand this to `if self.category == Category.OTHER`?

File: pyhap/accessory.py [hunk 4]
```diff
@@ -273,6 +290,43 @@ def remove_paired_client(self, client_uuid):
...
+            buffer[4] = buffer[4] | 1 << 7
+
+        value_high = self.category >> 1
+        struct.pack_into('>L', buffer, 0, value_high)
+
+        encoded_payload = base36.dumps(struct.unpack_from('>L', buffer, 4)[0]
+                                       + (struct.unpack_from('>L', buffer, 0)[0] * (2**32))).upper()
+        while len(encoded_payload) < 9:
```
Comment (ikalchev) [inline pyhap/accessory.py]: Better: `encoded_payload.rjust(9, "0")`

File: pyhap/accessory.py [hunk 5]
```diff
@@ -273,6 +290,43 @@ def remove_paired_client(self, client_uuid):
...
+            encoded_payload = '0' + encoded_payload
+
+        return 'X-HM://' + encoded_payload + self.setup_id
+
+    @property
+    def qr_code(self):
+        """Generates a pyqrcode object which can be used to store SVG, or print to
```
Comment (ikalchev) [inline pyhap/accessory.py]: How about: Generate a QR code for pairing with this accessory.

File: pyhap/accessory_driver.py [hunk 6]
```diff
@@ -63,6 +64,13 @@ def __init__(self, accessory, address, port):
              adv_data,
              pubname)
 
+    @property
```
Comment (ikalchev) [inline pyhap/accessory_driver.py]: I would leave this as a method - my understanding is that `@property` should be for simple computations. Anyways, I don't mind.

File: pyhap/accessory.py (hunk 5 above)
Comment (jslay88) [reply pyhap/accessory.py]: No problem.

File: pyhap/accessory.py (hunk 4 above)
Comment (jslay88) [reply pyhap/accessory.py]: Done in next commit.

File: pyhap/accessory.py (hunk 3 above)
Comment (jslay88) [reply pyhap/accessory.py]: Done in next commit.

File: pyhap/accessory.py (hunk 2 above)
Comment (jslay88) [reply pyhap/accessory.py]: You're right. I was just following the HAP-NodeJS, not thinking.

File: pyhap/accessory.py (hunk 1 above)
Comment (jslay88) [reply pyhap/accessory.py]: Yup, didn't know you could do the `or` in there like that. Learned something new!

Left unprotected, as technically, the accessory driver will need it for setup hash.

File: pyhap/accessory.py [hunk 7]
```diff
@@ -114,7 +118,7 @@ def create(cls, display_name, pincode, aid=STANDALONE_AID):
         return cls(display_name, aid=aid, mac=mac, pincode=pincode)
 
     def __init__(self, display_name, aid=None, mac=None, pincode=None,
-                 iid_manager=None):
+                 iid_manager=None, setup_id=None):
```
[the remark this answers is not in the document]
Comment (jslay88) [reply pyhap/accessory.py]: Done in next commit.

File: pyhap/accessory.py [hunk 8]
```diff
@@ -2,6 +2,10 @@
 import threading
 import logging
 import itertools
+import struct
+import base36
```
[the remark this answers is not in the document]
Comment (jslay88) [reply pyhap/accessory.py]: Done in next commit.

File: main.py [hunk 9]
```diff
@@ -40,6 +40,8 @@ def get_accessory():
 
 # Start the accessory on port 51826
 driver = AccessoryDriver(acc, port=51826)
+# Scan this QR code with your iOS device.
+acc.print_qr()
```
[the remark this answers is not in the document]
Comment (jslay88) [reply main.py]: Do we want to print, or use logger? Will logger display correctly?

File: pyhap/accessory_driver.py (hunk 6 above)
Comment (jslay88) [reply pyhap/accessory_driver.py]: Done in next commit.

File: main.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ikalchev) [reply main.py]: Print works I guess, as logger can be set to a file.

File: pyhap/accessory.py (hunk 2 above)
Comment (ikalchev) [reply pyhap/accessory.py]: totally fine, no need to reinvent the wheel. And that's the purpose of the review.

**Comment by jslay88** [comment]
Please see commit for the changes.

File: pyhap/accessory.py [hunk 10]
```diff
@@ -273,6 +292,42 @@ def remove_paired_client(self, client_uuid):
...
+        struct.pack_into('>L', buffer, 4, value_low)
+
+        if self.category == Category.OTHER:
+            buffer[4] = buffer[4] | 1 << 7
+
+        value_high = self.category >> 1
+        struct.pack_into('>L', buffer, 0, value_high)
+
+        encoded_payload = base36.dumps(struct.unpack_from('>L', buffer, 4)[0]
+                                       + (struct.unpack_from('>L', buffer, 0)[0] * (2**32))).upper()
```
Comment (ikalchev) [inline pyhap/accessory.py]: Instead of calculating 2 * 32 you can just shift left:
```python
>>> 1 << 32
4294967296
>>> 2 ** 32
4294967296
```

**Comment by jslay88** [comment]
Please see commit.

**Comment by jslay88** [comment]
Please see commit 51311fc

File: pyhap/accessory.py [hunk 11]
```diff
@@ -135,12 +139,18 @@ def __init__(self, display_name, aid=None, mac=None, pincode=None,
...
+            every time the instance is started. If not provided on init, will be random.
+            4 digit string 0-9 A-Z
+        @type setup_id: str
         """
         self.display_name = display_name
         self.aid = aid
         self.mac = mac
         self.config_version = 2
         self.reachable = True
+        self.pincode = pincode
```
Comment (ikalchev) [inline pyhap/accessory.py]: This should be removed: see the pincode property/method

Comment (jslay88) [reply pyhap/accessory.py]: ahh, sorry. missed it.

**Comment by jslay88** [comment]
Please see commit 67eb48d

Maybe finally :)

File: pyhap/accessory.py (hunk 11 above)
Comment (jslay88) [reply pyhap/accessory.py]: ahh sorry, missed it.

**Pull request merged**